appvalidate/palmtree_secrets.go

```diff
@@ -0,0 +1,181 @@
+package appvalidate
+
+import (
+	"bytes"
+	"context"
+	"crypto/tls"
+	"encoding/json"
+	"errors"
+	"fmt"
+	"net/http"
+	"net/url"
+	"path"
+
+	"github.com/kelseyhightower/envconfig"
+
+	"github.com/tidepool-org/platform/structure"
+	structValidator "github.com/tidepool-org/platform/structure/validator"
+)
+
+var (
+	ErrPalmTreeEmptyConfig = errors.New("empty PalmTree config")
+
+	partners = []string{
+		PartnerCoastal,
+		PartnerPalmTree,
+	}
+)
+
+const (
+	PartnerPalmTree = "PalmTree"
+)
+
+type PalmTreeSecretsConfig struct {
+	BaseURL   string `envconfig:"PALMTREE_BASE_URL"`
+	CalID     string `envconfig:"PALMTREE_CAL_ID"`
+	ProfileID string `envconfig:"PALMTREE_PROFILE_ID"`
+	// CertData is the raw contents of the tls certificate file
+	CertData []byte `envconfig:"PALMTREE_TLS_CERT_DATA"`
+	// KeyData is the raw contents of the tls private key file
+	KeyData []byte `envconfig:"PALMTREE_TLS_KEY_DATA"`
+}
+
+type PalmTreeSecrets struct {
+	Config PalmTreeSecretsConfig
+	client *http.Client
+}
+
+func NewPalmTreeSecrets() (*PalmTreeSecrets, error) {
+	cfg, err := newPalmTreeSecretsConfig()
```

Comment:
if you were to pass the cfg as an argument, this function wouldn't even be necessary, but if you wanted to keep it anyway (sometimes its nice to have a constuctor), it'd be simpler,as it wouldn't return an error. Plus then tests would be easier to write, as the config provided could be easily faked/mocked/whatever.

Why aren't there any tests?

Reply:
Not sure the value of tests here as it's only calling out to external tests. The only other thing would be validating the certificates but those are also just standard library calls. Any tests in particular you have in mind?

Reply:
Tests help other people understand what the code is doing. They can detect regressions, and generally improve the design of code. 

It'd be interesting to see some tests around `b64StdEncodingToURLEncoding`, because I'm not sure it's doing what it's supposed to, a test would give me confidence that it's doing the right things.

The code that handles continuing with invalid partner secrets would be another place that it would be good to have confidence that it can continue to function as expected without validated secret data.